### PR TITLE
Run the pr-label check on PR closed action and validate closed_by

### LINKED ISSUE #68459: Soft mandatory labeling for release notes
Composing release notes has historically been a very time consuming process. Until now, we would have one dedicated engineer from the PyTorch Core team manually categorize all the commits (usually about 3-3.5k commits)  every release, which is usually about a week's worth of full-time work. 

In PyTorch 1.10, we made a significant improvement, by auto-categorizing about 30% (1k) of the 3.5k commits using [this](https://github.com/pytorch/pytorch/pull/64929) script that labeled PRs based on the files changed. 

Now, we are proposing to move to an upfront labeling system, where all PR authors categorize and topic each PR upfront, which will make compiling release notes much easier at the time of release with very little upfront work. This would mean that in the future releases, categorizing of PRs for compiling release notes will be solely done based on the category labels. 

Inspired by https://github.com/pytorch/vision/pull/4618 we propose to add GitHub workflow/Probot path that would remind authors to assign *one* of the following category labels to PRs:

- `release notes: mobile`
- `release notes: jit`
- `release notes: visualization`
- `release notes: onnx`
- `release notes: quantization`
- `release notes: rocm`
- `release notes: cuda`
- `release notes: benchmark`
- `release notes: profiler`
- `release notes: performance_as_product`
- `release notes: package/deploy`
- `release notes: composability`
- `release notes: releng`
- `release notes: fx`
- `release notes: vulkan`
- `release notes: nn`
- `release notes: cpp`
- `release notes notes: python_frontend`
- `release notes: complex`
- `release notes: autograd`
- `release notes: build`
- `release notes: memory format`
- `release notes: foreach_frontend`
- `release notes: dataloader`
- `release notes: linalg_frontend`
- `release notes: Meta API`
- `release notes: sparse`
- `release notes: distributed (c10d)`
- `release notes: distributed (rpc)`
- `release notes: distributed (ddp)`
- `release notes: distributed (pipeline)`
- `release notes: distributed (fsdp)`
- `release notes: distributed (sharded)`

Please leave a comment, with additional category labels if you are working on a new feature or want to refine an existing category label into more specific category labels to distribute the work or for other reasons. Example., at the time of writing this, `distributed` category is split into six categories (`c10d`, `rpc`, `ddp`, `pipeline`, `fsdp`, `sharded`)

And optionally, assign *one* of the following topic labels:

- `topic: bc_breaking`
- `topic: deprecation`
- `topic: new feature`
- `topic: improvements`
- `topic: bug fixes`
- `topic: performance`
- `topic: documentation`
- `topic: developer feature`
- `topic: non-user visible` <- assign this label if you think PR should not be mentioned in release notes

Later assigning such labels might become a mandatory pre-merge check

cc @brianjo @mruberry @seemethere @malfet @pytorch/pytorch-dev-infra

### PR DESCRIPTION
**Pull request by atalman**
Fixes #68459
This PR implements reminder to assign the release notes and topic labels to each PR when merged. Here is an example of the message that set on the issue related to PR: 

> Hey atalman. You merged this PR, but no release notes category and topic labels were added. >
> The list of valid release and topic labels is available https://github.com/pytorch/pytorch/labels?q=release+notes+or+topic

Tested by manually running process_commit.py script in standalone mode passing following  commit_hash = "e020414cb25cd763103f77a10c6225ce27cbbb6e" which should resolve to this PR

### CODE REVIEW

**Review by bigfootjon — APPROVED** [review]
LGTM

**Review by bigfootjon — CHANGES_REQUESTED** [review]
Wait, this won't work for the `fbsync` branch either.

ShipIt sync to `fbsync` does not merge PRs, so this code will fail:

https://github.com/pytorch/pytorch/blob/master/.github/scripts/process_commit.py#L36-L41

Instead, you need to parse the commit message and extract the "Pull Request Resolved:" line with something like this:

https://github.com/pytorch/pytorch/blob/master/.github/scripts/process_commit.py#L36-L41

CC @malfet

**Comment by atalman** [comment]
> Wait, this won't work for the `fbsync` branch either.
> 
> ShipIt sync to `fbsync` does not merge PRs, so this code will fail:
> 
> https://github.com/pytorch/pytorch/blob/master/.github/scripts/process_commit.py#L36-L41
> 
> Instead, you need to parse the commit message and extract the "Pull Request Resolved:" line with something like this:
> 
> https://github.com/pytorch/pytorch/blob/master/.github/scripts/process_commit.py#L36-L41
> 
> CC @malfet

As discussed over with @bigfootjon  Implemented getting issue information and closed_by information and checking for that.

File: .github/scripts/process_commit.py [hunk 1]
```diff
@@ -1,10 +1,10 @@
 #!/usr/bin/env python3
 """
-This script finds the merger responsible for labeling a PR by a commit SHA. It is used by the workflow in
+This script finds the user/pr creator responsible for labeling a PR by a commit SHA. It is used by the workflow in
 '.github/workflows/pr-labels.yml'. If there exists no PR associated with the commit or the PR is properly labeled,
 this script is a no-op.
 
-Note: we ping the merger only, not the reviewers, as the reviewers can sometimes be external to pytorch
+Note: we ping the user only, not the reviewers, as the reviewers can sometimes be external to pytorch
```
Comment (malfet) [inline .github/scripts/process_commit.py:7]: s/merger/author?
```suggestion
Note: we ping the author only, not the reviewers, as the reviewers can sometimes be external to pytorch
```

File: .github/scripts/process_commit.py [hunk 2]
```diff
@@ -41,26 +42,32 @@ def get_pr_number(commit_hash: str) -> Any:
     return data[0]["number"]
 
 
-def get_pr_merger_and_labels(pr_number: int) -> Tuple[str, Set[str]]:
+def get_pr_user_and_labels(pr_number: int) -> Tuple[str, Set[str]]:
```
Comment (malfet) [inline .github/scripts/process_commit.py:45]:
```suggestion
def get_pr_author_and_labels(pr_number: int) -> Tuple[str, Set[str]]:
```

**Review by malfet — APPROVED** [review]
Looks good to me, but why not run this one on pushes to master and then extract "Pull request resolved" from the commit description?

Extracting comment is more complicated, its not directly attached to original PR that we are triggering at, we will have to do more requests to find out the commit that was made to master by pytorchboot . And then parse this commit. I think the fact that: PR against master was closed and issue got merged by pytorchboot is a good indication that we should run the logic. I don't think we can extract it on pushes to master since we want to attach warning comment to original PR and notify the owner of original PR. Hence original PR should be our starting and ending point.

File: .github/scripts/process_commit.py [hunk 3]
```diff
@@ -41,26 +42,32 @@ def get_pr_number(commit_hash: str) -> Any:
...
     commit_hash = sys.argv[1]
     pr_number = get_pr_number(commit_hash)
+    merger = get_pr_closed_by_name(pr_number)
 
-    if not pr_number:
+    if not pr_number or (merger != BOT_MERGER_NAME):
         sys.exit(0)
 
-    merger, labels = get_pr_merger_and_labels(pr_number)
+    user, labels = get_pr_user_and_labels(pr_number)
```
Comment (atalman) [inline .github/scripts/process_commit.py:66]:
```suggestion
    user, labels = get_pr_author_and_labels(pr_number)
```

File: .github/workflows/pr-labels.yml [hunk 4]
```diff
@@ -1,9 +1,15 @@
...
-      - master
+pull_request:
+  types: [closed]
+  branches:
+    - master
+
+env:
+  concurrency:
+  group: pr-labels-${{ github.event.pull_request.number || github.sha }}-${{ github.event_name == 'workflow_dispatch' }}
+  cancel-in-progress: true
```
Comment (seemethere) [inline .github/workflows/pr-labels.yml:12]:
```suggestion
concurrency:
  group: pr-labels-${{ github.event.pull_request.number || github.sha }}-${{ github.event_name == 'workflow_dispatch' }}
  cancel-in-progress: true
```

**Review by seemethere — COMMENTED** [review]
`concurrency:` is a top level directive

File: .github/scripts/process_commit.py [hunk 5]
```diff
@@ -15,6 +15,7 @@
 import requests
 
 # For a PR to be properly labeled it should have release notes label and one topic label
+BOT_MERGER_NAME = "pytorchmergebot"
 PRIMARY_LEBEL_FILTER = "release notes:"
 SECONDARY_LEBELS = {
```
Comment (janeyx99) [inline .github/scripts/process_commit.py:20]:
```suggestion
SECONDARY_LABELS = {
```

File: .github/scripts/process_commit.py [hunk 6]
```diff
@@ -15,6 +15,7 @@
 import requests
 
 # For a PR to be properly labeled it should have release notes label and one topic label
+BOT_MERGER_NAME = "pytorchmergebot"
 PRIMARY_LEBEL_FILTER = "release notes:"
```
Comment (janeyx99) [inline .github/scripts/process_commit.py:19]:
```suggestion
PRIMARY_LABEL_FILTER = "release notes:"
```
nit

File: .github/scripts/process_commit.py (hunk 5 above)
Comment (janeyx99) [reply .github/scripts/process_commit.py:20]: would require replacing other references

**Comment by malfet** [comment]
This PR will be triggered on PRs targeting other branches (release for example)
On the other hand, we want those to be triggered for ghstack PRs

**Comment by atalman** [comment]
Hey atalman. You merged this PR, but no release notes category and topic labels were added.     The list of valid release and topic labels is available https://github.com/pytorch/pytorch/labels?page=2&q=release+notes+or+topic

**Comment by malfet** [comment]
Can you please add a more reasonable description to PR?

File: .github/scripts/process_commit.py [hunk 7]
```diff
@@ -27,27 +29,43 @@
...
+    user = data["user"]["login"]
     labels = {label["name"] for label in data["labels"]}
-    return merger, labels
+    return user, labels
 
+def post_pytroch_comment(pr_number: int, merger: str) -> Any:
+    message = {'body' : f"Hey {merger}. \
+    You merged this PR, but no release notes category and topic labels were added. \
+    The list of valid release and topic labels is available \
+    https://github.com/pytorch/pytorch/labels?page=2&q=release+notes+or+topic"}
```
Comment (malfet) [inline .github/scripts/process_commit.py:62]: Why linking to page2?

File: .github/scripts/process_commit.py [hunk 8]
```diff
@@ -27,27 +29,43 @@
...
+def get_pr_user_and_labels(pr_number: int) -> Tuple[str, Set[str]]:
     # See https://docs.github.com/en/rest/reference/pulls#get-a-pull-request
     data = query_pytroch(f"pulls/{pr_number}", accept="application/vnd.github.v3+json")
-    merger = data["merged_by"]["login"]
+    user = data["user"]["login"]
     labels = {label["name"] for label in data["labels"]}
-    return merger, labels
+    return user, labels
 
+def post_pytroch_comment(pr_number: int, merger: str) -> Any:
```
Comment (malfet) [inline .github/scripts/process_commit.py:58]:
```suggestion
def post_pytorch_comment(pr_number: int, merger: str) -> Any:
```

File: .github/scripts/process_commit.py [hunk 9]
```diff
@@ -56,11 +74,17 @@ def get_pr_merger_and_labels(pr_number: int) -> Tuple[str, Set[str]]:
...
-    response_labels = list(map(lambda x: str(x["name"]), response.json()))
-    primary_labels = set(filter(lambda x: x.startswith(PRIMARY_LEBEL_FILTER), response_labels))
+    user, labels = get_pr_user_and_labels(pr_number)
+    collected_labels: List[str] = list()
+    for page in range(0, 10):
+        response = query_pytroch(f"labels?per_page=100&page={page}", accept="application/json")
+        page_labels = list(map(lambda x: str(x["name"]), response))
+        if not page_labels:
+            break
+            collected_labels += page_labels
```
Comment (malfet) [inline .github/scripts/process_commit.py:84]: Please move it to a separate function(say `collect_pytorch_labels`)

**Review by malfet — CHANGES_REQUESTED** [review]
Please fix typos in the code as well as add more detailed PR description

File: .github/scripts/process_commit.py (hunk 2 above)
Comment (atalman) [reply .github/scripts/process_commit.py:51]: This terminology comes from Github API. Reference:  https://docs.github.com/en/rest/reference/pulls#list-pull-requests--code-samples . Hence kept here as well for consistency. But true we are actually looking for authon.

File: .github/scripts/process_commit.py [hunk 10]
```diff
@@ -56,11 +84,11 @@ def get_pr_merger_and_labels(pr_number: int) -> Tuple[str, Set[str]]:
...
 
-    merger, labels = get_pr_merger_and_labels(pr_number)
-    response = query_pytroch("labels", accept="application/json")
-    response_labels = list(map(lambda x: str(x["name"]), response.json()))
-    primary_labels = set(filter(lambda x: x.startswith(PRIMARY_LEBEL_FILTER), response_labels))
-    is_properly_labeled = bool(primary_labels.intersection(labels) and SECONDARY_LEBELS.intersection(labels))
+    user, labels = get_pr_author_and_labels(pr_number)
+    repo_labels = get_repo_labels()
+
+    primary_labels = set(filter(lambda x: x.startswith(PRIMARY_LABEL_FILTER), repo_labels))
```
Comment (malfet) [inline .github/scripts/process_commit.py:90]: Why not extract secondary labels in the same fashion rather than have a hardcoded list?

**Review by malfet — APPROVED** [review]

File: .github/scripts/process_commit.py (hunk 10 above)
Comment (atalman) [reply .github/scripts/process_commit.py:90]: True, this is what I initially had in this PR: https://github.com/pytorch/pytorch/pull/71726
However as per discussion it was decided to have the secondary labels be hardcoded

**Comment by atalman** [comment]
Hey atalman.     NEW Message You merged this PR, but no release notes category and topic labels were added.     The list of valid release and topic labels is available     https://github.com/pytorch/pytorch/labels?q=release+notes+or+topic

**Comment by atalman** [comment]
> ey atalman. NEW Mess

For some reason, this does not work in GHA but works locally this is debug message

**Comment by atalman** [comment]
Hey atalman.     You merged this PR, but no release notes category and topic labels were added.     The list of valid release and topic labels is available     https://github.com/pytorch/pytorch/labels?q=release+notes+or+topic
